### ISSUE TITLE
Tuya ME-81H thermostat problem

### ISSUE THREAD

**Issue by bearpawmaxim** [esphome/issues#3270]
### The problem

Hi all!
Have configured tuya climate integration for ME-81H thermostat but it behaves abnormally:
1. No datapoints reported in log after "If no further output is received, confirm that this is a supported Tuya device." message.
2. The thermostat itself does not shows the WiFi icon on the screen.
3. The time is not synchronized.
4. When turning on the thermostat, it reports it's status as OFF. After toggling the mode through HA everything becomes normal.

Despite all these errors i can control the target temperature and lock/unlock the thermostat.
Datapoint IDs are obtained from WThermostatBeca, with which it successfully worked before.

### Which version of ESPHome has the issue?

2022.4.0

### What type of installation are you using?

Home Assistant Add-on

### Which version of Home Assistant has the issue?

2022.5.2

### What platform are you using?

ESP8266

### Board

TYWE3S

### Component causing the issue

tuya

### Example YAML snippet

```yaml
substitutions:
  device_name: warm-floor

packages:
  device_common: !include .device_common.yaml

esphome:
  name: ${device_name}
  platform: ESP8266
  board: esp01_1m

logger:
  baud_rate: 0

uart:
  rx_pin: GPIO3
  tx_pin: GPIO1
  baud_rate: 9600

tuya:
  time_id: sntp_time

climate:
  - platform: tuya
    name: "Warm floor"
    switch_datapoint: 1
    target_temperature_datapoint: 16
    target_temperature_multiplier: 1.0
    current_temperature_datapoint: 24
    current_temperature_multiplier: 0.1

switch:
  - platform: "tuya"
    name: "Warm floor lock"
    switch_datapoint: 40
```

### Anything in the logs that might be useful for us?

```txt
INFO Reading configuration /config/esphome/warm-floor.yaml...
INFO Starting log output from warm-floor.local using esphome API
INFO Successfully connected to warm-floor.local
[22:29:35][I][app:102]: ESPHome version 2022.4.0 compiled on May  8 2022, 22:11:54
[22:29:35][C][wifi:491]: WiFi:
[22:29:35][C][wifi:353]:   Local MAC: ********************
[22:29:35][C][wifi:354]:   SSID: *****************
[22:29:35][C][wifi:355]:   IP Address: ************
[22:29:35][C][wifi:356]:   BSSID: ***************************
[22:29:35][C][wifi:358]:   Hostname: 'warm-floor'
[22:29:35][C][wifi:360]:   Signal strength: -61 dB ▂▄▆█
[22:29:35][C][wifi:364]:   Channel: 11
[22:29:35][C][wifi:365]:   Subnet: 255.255.255.0
[22:29:35][C][wifi:366]:   Gateway: 192.168.1.1
[22:29:35][C][wifi:367]:   DNS1: 192.168.1.1
[22:29:35][C][wifi:368]:   DNS2: 0.0.0.0
[22:29:35][C][logger:233]: Logger:
[22:29:35][C][logger:234]:   Level: DEBUG
[22:29:35][C][logger:235]:   Log Baud Rate: 0
[22:29:35][C][logger:236]:   Hardware UART: UART0
[22:29:35][C][uart.arduino_esp8266:102]: UART Bus:
[22:29:35][C][uart.arduino_esp8266:103]:   TX Pin: GPIO1
[22:29:35][C][uart.arduino_esp8266:104]:   RX Pin: GPIO3
[22:29:35][C][uart.arduino_esp8266:106]:   RX Buffer Size: 256
[22:29:35][C][uart.arduino_esp8266:108]:   Baud Rate: 9600 baud
[22:29:35][C][uart.arduino_esp8266:109]:   Data Bits: 8
[22:29:35][C][uart.arduino_esp8266:110]:   Parity: NONE
[22:29:35][C][uart.arduino_esp8266:111]:   Stop bits: 1
[22:29:35][C][uart.arduino_esp8266:113]:   Using hardware serial interface.
[22:29:35][C][uptime.sensor:031]: Uptime Sensor 'warm-floor uptime seconds'
[22:29:35][C][uptime.sensor:031]:   State Class: 'total_increasing'
[22:29:35][C][uptime.sensor:031]:   Unit of Measurement: 's'
[22:29:35][C][uptime.sensor:031]:   Accuracy Decimals: 0
[22:29:35][C][uptime.sensor:031]:   Icon: 'mdi:timer-outline'
[22:29:35][C][template.text_sensor:021]: Template Sensor 'warm-floor uptime'
[22:29:35][C][template.text_sensor:021]:   Icon: 'mdi:clock-start'
[22:29:35][C][tuya.climate:140]: Tuya Climate 'Warm floor'
[22:29:35][C][tuya.climate:142]:   Switch has datapoint ID 1
[22:29:35][C][tuya.climate:146]:   Target Temperature has datapoint ID 16
[22:29:35][C][tuya.climate:148]:   Current Temperature has datapoint ID 24
[22:29:35][C][tuya.switch:023]: Tuya Switch 'Warm floor lock'
[22:29:35][C][tuya.switch:024]:   Switch has datapoint ID 40
[22:29:35][C][captive_portal:088]: Captive Portal:
[22:29:35][C][mdns:084]: mDNS:
[22:29:35][C][mdns:085]:   Hostname: warm-floor
[22:29:35][C][ota:085]: Over-The-Air Updates:
[22:29:35][C][ota:086]:   Address: warm-floor.local:8266
[22:29:35][C][ota:089]:   Using Password.
[22:29:35][C][api:138]: API Server:
[22:29:35][C][api:139]:   Address: warm-floor.local:6053
[22:29:35][C][api:143]:   Using noise encryption: NO
[22:29:35][C][sntp:050]: SNTP Time:
[22:29:35][C][sntp:051]:   Server 1: '0.pool.ntp.org'
[22:29:35][C][sntp:052]:   Server 2: '1.pool.ntp.org'
[22:29:35][C][sntp:053]:   Server 3: '2.pool.ntp.org'
[22:29:35][C][sntp:054]:   Timezone: 'EET-2EEST,M3.5.0/3,M10.5.0/4'
[22:29:35][C][wifi_signal.sensor:009]: WiFi Signal 'warm-floor WiFi Signal'
[22:29:35][C][wifi_signal.sensor:009]:   Device Class: 'signal_strength'
[22:29:35][C][wifi_signal.sensor:009]:   State Class: 'measurement'
[22:29:35][C][wifi_signal.sensor:009]:   Unit of Measurement: 'dBm'
[22:29:35][C][wifi_signal.sensor:009]:   Accuracy Decimals: 0
[22:29:35][C][tuya:028]: Tuya:
[22:29:35][C][tuya:030]:   Configuration will be reported when setup is complete. Current init_state: 1
[22:29:35][C][tuya:032]:   If no further output is received, confirm that this is a supported Tuya device.
[22:29:48][D][tuya:275]: Datapoint 36 update to 1
[22:29:49][D][tuya:263]: Datapoint 24 update to 220
[22:29:49][D][climate:385]: 'Warm floor' - Sending state:
[22:29:49][D][climate:388]:   Mode: HEAT
[22:29:49][D][climate:390]:   Action: IDLE
[22:29:49][D][climate:408]:   Current Temperature: 22.00°C
[22:29:49][D][climate:414]:   Target Temperature: 22.00°C
[22:29:49][D][tuya:292]: Datapoint 45 update to 00000000
[22:30:04][D][tuya:275]: Datapoint 36 update to 1
[22:30:04][D][tuya:263]: Datapoint 24 update to 220
[22:30:04][D][climate:385]: 'Warm floor' - Sending state:
[22:30:04][D][climate:388]:   Mode: HEAT
[22:30:04][D][climate:390]:   Action: IDLE
[22:30:04][D][climate:408]:   Current Temperature: 22.00°C
[22:30:04][D][climate:414]:   Target Temperature: 22.00°C
[22:30:04][D][tuya:292]: Datapoint 45 update to 00000000
[22:30:05][D][sensor:124]: 'warm-floor WiFi Signal': Sending state -57.00000 dBm with 0 decimals of accuracy
[22:30:09][D][sensor:124]: 'warm-floor uptime seconds': Sending state 1006.13300 s with 0 decimals of accuracy
```

**Comment by bearpawmaxim** [comment]
UPD:
Set the logging level to VERBOSE and got those logs
`INFO Reading configuration /config/esphome/warm-floor.yaml...
INFO Starting log output from warm-floor.local using esphome API
INFO Successfully connected to warm-floor.local
[08:34:41][I][app:102]: ESPHome version 2022.4.0 compiled on May  9 2022, 00:32:29
[08:34:41][C][wifi:491]: WiFi:
[08:34:41][C][wifi:353]:   Local MAC: ***
[08:34:41][C][wifi:354]:   SSID: ***
[08:34:41][C][wifi:355]:   IP Address: ***
[08:34:41][C][wifi:356]:   BSSID: ***
[08:34:41][C][wifi:358]:   Hostname: 'warm-floor'
[08:34:41][C][wifi:360]:   Signal strength: -62 dB в–‚в–„в–†в–€
[08:34:41][C][wifi:364]:   Channel: 11
[08:34:41][C][wifi:365]:   Subnet: 255.255.255.0
[08:34:41][C][wifi:366]:   Gateway: 192.168.1.1
[08:34:41][C][wifi:367]:   DNS1: 192.168.1.1
[08:34:41][C][wifi:368]:   DNS2: 0.0.0.0
[08:34:41][C][logger:233]: Logger:
[08:34:41][C][logger:234]:   Level: VERBOSE
[08:34:41][C][logger:235]:   Log Baud Rate: 0
[08:34:41][C][logger:236]:   Hardware UART: UART0
[08:34:41][C][uart.arduino_esp8266:102]: UART Bus:
[08:34:41][C][uart.arduino_esp8266:103]:   TX Pin: GPIO1
[08:34:41][C][uart.arduino_esp8266:104]:   RX Pin: GPIO3
[08:34:41][C][uart.arduino_esp8266:106]:   RX Buffer Size: 256
[08:34:41][C][uart.arduino_esp8266:108]:   Baud Rate: 9600 baud
[08:34:41][C][uart.arduino_esp8266:109]:   Data Bits: 8
[08:34:41][C][uart.arduino_esp8266:110]:   Parity: NONE
[08:34:41][C][uart.arduino_esp8266:111]:   Stop bits: 1
[08:34:41][C][uart.arduino_esp8266:113]:   Using hardware serial interface.
[08:34:41][C][uptime.sensor:031]: Uptime Sensor 'warm-floor uptime seconds'
[08:34:41][C][uptime.sensor:031]:   State Class: 'total_increasing'
[08:34:41][C][uptime.sensor:031]:   Unit of Measurement: 's'
[08:34:41][C][uptime.sensor:031]:   Accuracy Decimals: 0
[08:34:41][C][uptime.sensor:031]:   Icon: 'mdi:timer-outline'
[08:34:41][V][uptime.sensor:031]:   Unique ID: 'e8db84bb3932-uptime'
[08:34:41][C][template.text_sensor:021]: Template Sensor 'warm-floor uptime'
[08:34:41][C][template.text_sensor:021]:   Icon: 'mdi:clock-start'
[08:34:41][C][tuya.climate:140]: Tuya Climate 'Warm floor'
[08:34:41][C][tuya.climate:142]:   Switch has datapoint ID 1
[08:34:41][C][tuya.climate:146]:   Target Temperature has datapoint ID 16
[08:34:41][C][tuya.climate:148]:   Current Temperature has datapoint ID 24
[08:34:41][C][tuya.switch:023]: Tuya Switch 'Warm floor lock'
[08:34:41][C][tuya.switch:024]:   Switch has datapoint ID 40
[08:34:41][C][captive_portal:088]: Captive Portal:
[08:34:41][C][mdns:084]: mDNS:
[08:34:41][C][mdns:085]:   Hostname: warm-floor
[08:34:41][V][mdns:086]:   Services:
[08:34:41][V][mdns:088]:   - _esphomelib, _tcp, 6053
[08:34:41][V][mdns:090]:     TXT: version = 2022.4.0
[08:34:41][V][mdns:090]:     TXT: mac = e8db84bb3932
[08:34:41][V][mdns:090]:     TXT: platform = ESP8266
[08:34:41][V][mdns:090]:     TXT: board = esp01_1m
[08:34:41][C][ota:085]: Over-The-Air Updates:
[08:34:41][C][ota:086]:   Address: warm-floor.local:8266
[08:34:41][C][ota:089]:   Using Password.
[08:34:41][C][api:138]: API Server:
[08:34:41][C][api:139]:   Address: warm-floor.local:6053
[08:34:41][C][api:143]:   Using noise encryption: NO
[08:34:41][C][sntp:050]: SNTP Time:
[08:34:41][C][sntp:051]:   Server 1: '0.pool.ntp.org'
[08:34:41][C][sntp:052]:   Server 2: '1.pool.ntp.org'
[08:34:41][C][sntp:053]:   Server 3: '2.pool.ntp.org'
[08:34:41][C][sntp:054]:   Timezone: 'EET-2EEST,M3.5.0/3,M10.5.0/4'
[08:34:41][C][wifi_signal.sensor:009]: WiFi Signal 'warm-floor WiFi Signal'
[08:34:41][C][wifi_signal.sensor:009]:   Device Class: 'signal_strength'
[08:34:41][C][wifi_signal.sensor:009]:   State Class: 'measurement'
[08:34:41][C][wifi_signal.sensor:009]:   Unit of Measurement: 'dBm'
[08:34:41][C][wifi_signal.sensor:009]:   Accuracy Decimals: 0
[08:34:41][V][wifi_signal.sensor:009]:   Unique ID: 'e8db84bb3932-wifisignal'
[08:34:41][C][tuya:028]: Tuya:
[08:34:41][C][tuya:030]:   Configuration will be reported when setup is complete. Current init_state: 1
[08:34:41][C][tuya:032]:   If no further output is received, confirm that this is a supported Tuya device.
[08:34:49][V][tuya:358]: Sending Tuya: CMD=0x00 VERSION=0 DATA=[] INIT_STATE=1
[08:34:49][V][tuya:108]: Received Tuya: CMD=0x00 VERSION=3 DATA=[01] INIT_STATE=1
[08:34:49][V][tuya:134]: MCU Heartbeat (0x01)
[08:34:55][V][tuya:108]: Received Tuya: CMD=0x07 VERSION=3 DATA=[24.04.00.01.01 (5)] INIT_STATE=1
[08:34:55][D][tuya:275]: Datapoint 36 update to 1
[08:34:55][V][tuya:108]: Received Tuya: CMD=0x07 VERSION=3 DATA=[18.02.00.04.00.00.00.E1 (8)] INIT_STATE=1
[08:34:55][D][tuya:263]: Datapoint 24 update to 225
[08:34:55][V][tuya.climate:056]: MCU reported current temperature is: 22.5
[08:34:55][D][climate:385]: 'Warm floor' - Sending state:
[08:34:55][D][climate:388]:   Mode: OFF
[08:34:55][D][climate:390]:   Action: OFF
[08:34:55][D][climate:408]:   Current Temperature: 22.50В°C
[08:34:55][D][climate:414]:   Target Temperature: 0.00В°C
[08:34:55][V][tuya:108]: Received Tuya: CMD=0x07 VERSION=3 DATA=[2D.05.00.01.00 (5)] INIT_STATE=1
[08:34:55][D][tuya:292]: Datapoint 45 update to 00000000
[08:35:04][V][tuya:358]: Sending Tuya: CMD=0x00 VERSION=0 DATA=[] INIT_STATE=1
[08:35:04][V][tuya:108]: Received Tuya: CMD=0x00 VERSION=3 DATA=[01] INIT_STATE=1
[08:35:04][V][tuya:134]: MCU Heartbeat (0x01)`

**Comment by bearpawmaxim** [comment]
Thanks to @ssieb, we found out that the MCU of my thermostat is laggy and very often loses the commands sent to it.
As a workaround, it was decided to retry command until we get the response.
The initial workaround is implemented in https://github.com/ssieb/esphome/commit/0136ac6ebe488bc85360132c676d2bc58494f457
With this workaround I've managed to complete the device initialization sequence(INIT_STATE=5) and to receive the datapoints dump.

Next problem was related to the wifi connection status (mcu wasn't able to know about network connection), that I've managed to fix in https://github.com/esphome/esphome/pull/3466
